<?xml version="1.0" encoding="utf-8"?>
<robot name="R5_DOF_Robot_Assembly">

  <!-- ===== world alignment helper ===== -->
  <link name="world"/>
  
 

  <!-- ===================== Base_Link ===================== -->
  <link name="Base_Link">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.8"/>
      <inertia ixx="0.002" ixy="0" ixz="0" iyy="0.002" iyz="0" izz="0.002"/>
    </inertial>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://robotics/meshes/Base_Link.STL"/>
      </geometry>
      <material name="red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 -0.030" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.060" length="0.020"/>
      </geometry>
    </collision>
  </link>

   <joint name="Base_Link_to_world" type="fixed">
    <parent link="world"/>
    <child link="Base_Link"/>
    <origin xyz="0 0 0.050" rpy="0 0 0"/>
  </joint>
  <!-- ===================== Rotating_Waste ===================== -->
  <link name="Rotating_Waste">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.7"/>
      <inertia ixx="0.0015" ixy="0" ixz="0" iyy="0.0015" iyz="0" izz="0.0015"/>
    </inertial>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://robotics/meshes/Rotating_Waste.STL"/>
      </geometry>
      <material name="gray">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.055" length="0.010"/>
      </geometry>
    </collision>
  </link>

  <joint name="Joint_1" type="revolute">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="Base_Link"/>
    <child link="Rotating_Waste"/>
    <axis xyz="0 0 1"/>
    <limit lower="-3.14159" upper="3.14159" effort="200" velocity="2.0"/>
    <dynamics damping="3.0" friction="0.2"/>
  </joint>

  <!-- ===================== Arm_1 ===================== -->
  <link name="Arm_1">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.6"/>
      <inertia ixx="0.0012" ixy="0" ixz="0" iyy="0.0004" iyz="0" izz="0.0012"/>
    </inertial>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://robotics/meshes/Arm_1.STL"/>
      </geometry>
      <material name="white">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 0.06" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.020" length="0.12"/>
      </geometry>
    </collision>
  </link>

  <joint name="Joint_2" type="revolute">
    <origin xyz="-2.7567E-05 0.0061999 0.04975" rpy="1.5708 0 0"/>
    <parent link="Rotating_Waste"/>
    <child link="Arm_1"/>
    <axis xyz="0 0 1"/>
    <limit lower="-3.14159" upper="3.14159" effort="200" velocity="2.0"/>
    <dynamics damping="3.0" friction="0.2"/>
  </joint>

  <!-- ===================== Arm_2 ===================== -->
  <link name="Arm_2">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.5"/>
      <inertia ixx="0.0009" ixy="0" ixz="0" iyy="0.0003" iyz="0" izz="0.0009"/>
    </inertial>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://robotics/meshes/Arm_2.STL"/>
      </geometry>z
      <material name="gray">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 0.06" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.018" length="0.12"/>
      </geometry>
    </collision>
  </link>

  <joint name="Joint_3" type="revolute">
    <origin xyz="0 0.140 0.000" rpy="0 0 1.5708"/>
    <parent link="Arm_1"/>
    <child link="Arm_2"/>
    <axis xyz="0 0 -1"/>
    <limit lower="-3.14159" upper="3.14159" effort="200" velocity="2.0"/>
    <dynamics damping="3.0" friction="0.2"/>
  </joint>

  <!-- ===================== Arm_3 ===================== -->
  <link name="Arm_3">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.4"/>
      <inertia ixx="0.0006" ixy="0" ixz="0" iyy="0.0002" iyz="0" izz="0.0006"/>
    </inertial>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://robotics/meshes/Arm_3.STL"/>
      </geometry>
      <material name="gray">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.015" length="0.140"/>
      </geometry>
    </collision>
  </link>

  <joint name="Joint_4" type="revolute">
    <origin xyz="0 0 0" rpy="-1.5708 0 -1.5708"/>
    <parent link="Arm_2"/>
    <child link="Arm_3"/>
    <axis xyz="0 0 -1"/>
    <limit lower="-3.14159" upper="3.14159" effort="200" velocity="2.0"/>
    <dynamics damping="3.0" friction="0.2"/>
  </joint>

  <!-- ===================== Gripper ===================== -->
  <link name="Gripper">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.3"/>
      <inertia ixx="0.0003" ixy="0" ixz="0" iyy="0.0001" iyz="0" izz="0.0003"/>
    </inertial>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://robotics/meshes/Gripper.STL"/>
      </geometry>
      <material name="red">
        <color rgba="0 0 0 0"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.030 0.080 0.030"/>
      </geometry>
    </collision>
  </link>

  <joint name="Joint_5" type="revolute">
    <origin xyz="-0.001651 -0.0054574 0.13289" rpy="1.5708 0 0"/>
    <parent link="Arm_3"/>
    <child link="Gripper"/>
    <axis xyz="0 0 1"/>
    <limit lower="-3.14159" upper="3.14159" effort="200" velocity="2.0"/>
    <dynamics damping="3.0" friction="0.2"/>
  </joint>

  <!-- ===================== End Effector Frame ===================== -->
  <link name="EE_frame">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.001"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.005"/>
      </geometry>
      <material name="green">
        <color rgba="0 1 0 1"/>
      </material>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.001"/>
      </geometry>
    </collision>
  </link>

  <joint name="EE_frame_lock" type="fixed">
    <parent link="Gripper"/>
    <child link="EE_frame"/>
    <origin xyz="0.007 0.1 0.013" rpy="0 0 0"/>
  </joint>

  <!-- ===================== ROS2 Control ===================== -->
  <ros2_control name="R5DOFController" type="system">
    <hardware>
      <plugin>gz_ros2_control/GazeboSimSystem</plugin>
    </hardware>
    <joint name="Joint_1">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="Joint_2">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="Joint_3">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="Joint_4">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="Joint_5">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
  </ros2_control>

  <gazebo>
    <plugin filename="gz_ros2_control-system" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
    <parameters>/home/dell/robotics/install/robotics/share/robotics/config/controllers.yaml</parameters>
    </plugin>
  </gazebo>

</robot>